client/blocks/gdpr-banner/index.jsx

```diff
@@ -29,83 +29,87 @@ const SIX_MONTHS = 6 * 30 * 24 * 60 * 60;
+		return false;
+	}
+	if ( isWpMobileApp() ) {
+		return false;
+	}
+	if ( isCurrentUserMaybeInGdprZone() ) {
+		return true;
+	}
+	return false;
+}
 
-	static defaultProps = {
-		recordCookieBannerOk: noop,
-		recordCookieBannerView: noop,
-		translate: identity,
-	};
+function GdprBanner( props ) {
+	const [ hideBanner, setHideBanner ] = useState( false );
+	const [ renderBanner, setRenderBanner ] = useState( false );
+	const translate = useTranslate();
 
-	state = {
-		showBanner: true,
-	};
+	const { recordCookieBannerOk, recordCookieBannerView } = props;
 
-	acknowledgeClicked = () => {
+	const acknowledgeClicked = useCallback( () => {
 		document.cookie = cookie.serialize( 'sensitive_pixel_option', 'yes', {
 			path: '/',
 			maxAge: SIX_MONTHS,
 		} );
-		this.props.recordCookieBannerOk();
-		this.setState( { showBanner: false } );
-	};
+		recordCookieBannerOk();
+		setHideBanner( true );
+	}, [ recordCookieBannerOk ] );
 
-	shouldShowBanner() {
-		// Don't render banner in SSR.
-		if ( ! hasDocument ) {
-			return false;
-		}
-		const cookies = cookie.parse( document.cookie );
-		if ( cookies.sensitive_pixel_option === 'yes' || cookies.sensitive_pixel_option === 'no' ) {
-			return false;
-		}
-		if ( isWpMobileApp() ) {
-			return false;
-		}
-		if ( isCurrentUserMaybeInGdprZone() ) {
-			return true;
-		}
-		return false;
+	// We want to ensure that the first render is always empty, to match the server.
+	// This avoids potential hydration issues.
+	useEffect( () => {
+		setRenderBanner( true );
+	}, [] );
+
+	if ( ! renderBanner || ! shouldShowBanner() ) {
+		return null;
 	}
 
-	render() {
-		if ( ! this.shouldShowBanner() ) {
-			return null;
+	const copy = translate(
+		'Our websites and dashboards use cookies. By continuing, you agree to their use. ' +
+			'{{a}}Learn more{{/a}}, including how to control cookies.',
+		{
+			components: {
+				a: <a href={ localizeUrl( 'https://automattic.com/cookies' ) } />,
+			},
 		}
-		const { translate } = this.props;
-		const copy = translate(
-			'Our websites and dashboards use cookies. By continuing, you agree to their use. ' +
-				'{{a}}Learn more{{/a}}, including how to control cookies.',
-			{
-				components: {
-					a: <a href={ localizeUrl( 'https://automattic.com/cookies' ) } />,
-				},
-			}
-		);
-		const classes = {
-			'is-compact': true,
-			'gdpr-banner': true,
-			'gdpr-banner__hiding': ! this.state.showBanner,
-		};
-		this.props.recordCookieBannerView();
-		return (
-			<Card className={ classNames( classes ) }>
-				<div className="gdpr-banner__text-content">{ preventWidows( decodeEntities( copy ) ) }</div>
-				<div className="gdpr-banner__buttons">
-					<Button className="gdpr-banner__acknowledge-button" onClick={ this.acknowledgeClicked }>
-						{ translate( 'Got it!' ) }
-					</Button>
-				</div>
-			</Card>
-		);
-	}
+	);
+	const classes = {
+		'is-compact': true,
+		'gdpr-banner': true,
+		'gdpr-banner__hiding': hideBanner,
+	};
+	recordCookieBannerView();
```

Comment:
This is a side-effect that shouldn't be in the render method. And it should be called only once, on mount, and only when the banner is actually displayed.
```js
useEffect( () => {
  showBanner && recordCookieBannerView();
}, [ showBanner ] );
```

Reply:
Ah, sorry, I'd missed this side-effect. The current logic already appears to account for when it's actually shown, but by extracting this into `useEffect` does need in-effect checking 👍

Reply:
It was a pre-existing bug: the view would be recorded multiple times (on every rerender), and Redux dispatch during React render could possibly have some evil consequences 🙂